File: sample-java/src/main/java/social/bigbone/sample/Authenticator.java
package social.bigbone.sample;

import com.google.gson.Gson;
import social.bigbone.MastodonClient;
import social.bigbone.MastodonRequest;
import social.bigbone.Parameters;
import social.bigbone.api.Scope;
import social.bigbone.api.entity.auth.AccessToken;
import social.bigbone.api.entity.auth.AppRegistration;
import social.bigbone.api.exception.BigBoneRequestException;
import social.bigbone.api.method.AppsMethods;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.Properties;

@SuppressWarnings("PMD.SystemPrintln")
final class Authenticator {
    private static final String CLIENT_ID = "client_id";
    private static final String CLIENT_SECRET = "client_secret";
    private static final String ACCESS_TOKEN = "access_token";

    private Authenticator() {}

    static MastodonClient appRegistrationIfNeeded(final String instanceName, final String credentialFilePath, final boolean useStreaming) throws IOException, BigBoneRequestException {
        final File file = new File(credentialFilePath);
        if (!file.exists()) {
            System.out.println("create $credentialFilePath.");
            file.createNewFile();
        }
        final Properties properties = new Properties();
        System.out.println("load $credentialFilePath.");
        properties.load(Files.newInputStream(file.toPath()));
        if (properties.get(CLIENT_ID) == null) {
            System.out.println("try app registration...");
            final AppRegistration appRegistration = appRegistration(instanceName);
            properties.put(CLIENT_ID, appRegistration.getClientId());
            properties.put(CLIENT_SECRET, appRegistration.getClientSecret());
            properties.store(Files.newOutputStream(file.toPath()), "app registration");
        } else {
            System.out.println("app registration found...");
        }
        final String clientId = properties.get(CLIENT_ID).toString();
        final String clientSecret = properties.get(CLIENT_SECRET).toString();
        if (properties.get(ACCESS_TOKEN) == null) {
            System.out.println("get access token for $instanceName...");
            System.out.println("please input your email...");
            final String email = System.console().readLine();
            System.out.println("please input your password...");
            final String pass = System.console().readLine();
            final AccessToken accessToken = getAccessToken(instanceName, clientId, clientSecret, email, pass);
            properties.put(ACCESS_TOKEN, accessToken.getAccessToken());
            properties.store(Files.newOutputStream(file.toPath()), "app registration");
        } else {
            System.out.println("access token found...");
        }
        final MastodonClient.Builder builder = new MastodonClient.Builder(instanceName)
            .accessToken(properties.get(ACCESS_TOKEN).toString());
        if (useStreaming) {
            builder.useStreamingApi();
        }
        return builder.build();
    }

    private static AccessToken getAccessToken(final String instanceName, final String clientId, final String clientSecret, final String email, final String password) throws BigBoneRequestException {
        final MastodonClient client = new MastodonClient.Builder(instanceName).build();
        return postUserNameAndPassword(client, clientId, clientSecret, new Scope(), email, password).execute();
    }

    private static AppRegistration appRegistration(final String instanceName) throws BigBoneRequestException {
        final MastodonClient client = new MastodonClient.Builder(instanceName).build();
        final AppsMethods appsMethods = new AppsMethods(client);
        return appsMethods.createApp("kotlindon", "urn:ietf:wg:oauth:2.0:oob", new Scope(), null).execute();
    }

    /**
     * Obtain an access token, to be used during API calls that are not public. This method uses a grant_type
     * that is undocumented in Mastodon API, and should NOT be used in production code. It will be removed at a later date.
     * Apps.getAccessToken() should be used instead.
     */
    private static MastodonRequest<AccessToken> postUserNameAndPassword(
            final MastodonClient client,
            final String clientId,
            final String clientSecret,
            final Scope scope,
            final String userName,
            final String password
            ) {
        final Parameters parameters = new Parameters()
                .append("client_id", clientId)
                .append("client_secret", clientSecret)
                .append("scope", scope.toString())
                .append("username", userName)
                .append("password", password)
                .append("grant_type", "password");
        return new MastodonRequest<>(
                () -> client.post("oauth/token", parameters),
                s -> new Gson().fromJson(s, AccessToken.class)
        );
    }
}

File: sample-kotlin/src/main/kotlin/social/bigbone/sample/Authenticator.kt
package social.bigbone.sample

import social.bigbone.MastodonClient
import social.bigbone.api.Scope
import social.bigbone.api.entity.Application
import social.bigbone.api.entity.Token
import social.bigbone.api.method.OAuthMethods
import java.io.File
import java.util.Properties

object Authenticator {
    private const val CLIENT_ID = "client_id"
    private const val CLIENT_SECRET = "client_secret"
    private const val ACCESS_TOKEN = "access_token"
    private const val REDIRECT_URI = "urn:ietf:wg:oauth:2.0:oob"

    // Request all non-admin scopes during app registration and while getting the user access token,
    // because individual samples might require any or all of them. Generally, an app should request
    // the minimum necessary for its intended functionality.
    private val fullScope = Scope(Scope.READ.ALL, Scope.WRITE.ALL, Scope.PUSH.ALL)

    fun appRegistrationIfNeeded(instanceName: String, credentialFilePath: String): MastodonClient {
        val file = File(credentialFilePath)
        if (!file.exists()) {
            println("create $credentialFilePath.")
            file.createNewFile()
        }
        val properties = Properties()
        println("load $credentialFilePath.")
        properties.load(file.inputStream())
        if (properties[CLIENT_ID] == null) {
            println("try app registration...")
            val appRegistration = appRegistration(instanceName)
            properties[CLIENT_ID] = appRegistration.clientId
            properties[CLIENT_SECRET] = appRegistration.clientSecret
            properties.store(file.outputStream(), "app registration")
        } else {
            println("app registration found...")
        }
        val clientId = properties[CLIENT_ID]?.toString() ?: error("client id not found")
        val clientSecret = properties[CLIENT_SECRET]?.toString() ?: error("client secret not found")

        if (properties[ACCESS_TOKEN] == null) {
            println("get access token for $instanceName...")
            println("please input your email...")
            val email = System.`in`.bufferedReader().readLine()
            println("please input your password...")
            val pass = System.`in`.bufferedReader().readLine()
            val accessToken = getUserAccessToken(
                instanceName,
                clientId,
                clientSecret,
                email,
                pass
            )
            properties[ACCESS_TOKEN] = accessToken.accessToken
            properties.store(file.outputStream(), "app registration")
        } else {
            println("access token found...")
        }
        return MastodonClient.Builder(instanceName)
            .accessToken(properties[ACCESS_TOKEN].toString())
            .build()
    }

    private fun getUserAccessToken(
        instanceName: String,
        clientId: String,
        clientSecret: String,
        email: String,
        password: String
    ): Token {
        val client = MastodonClient.Builder(instanceName).build()
        val oAuthMethods = OAuthMethods(client)
        return oAuthMethods.getUserAccessTokenWithPasswordGrant(clientId, clientSecret, REDIRECT_URI, email, password, fullScope).execute()
    }

    private fun appRegistration(instanceName: String): Application {
        val client = MastodonClient.Builder(instanceName).build()
        return client.apps.createApp(
            clientName = "bigbone-sample-app",
            redirectUris = REDIRECT_URI,
            scope = fullScope
        ).execute()
    }
}
